### ISSUE TITLE
mysql json field cannot update nested array item through nested array's index

### ISSUE THREAD

**Issue by eleven26** [laravel/framework#26415]
- Laravel Version: 5.7.12
- PHP Version: 7.1.14
- Database Driver & Version:  MySQL 5.7.17

### Description:
Laravel framework cannot update array item through array index like: `'$.test[0].name'`, the sql generate by `\Illuminate\Database\Query\Grammars\Grammar::wrapJsonPath` seem like this `'$."test[0]".name'`,  but with double quotation mark won't update MySQL json field,  at the same time, the update statement return the wrong result without updating MySQL's record, database record not updated but get the true result.

### source code
```
    /**
     * Wrap the given JSON path.
     *
     * @param  string  $value
     * @return string
     */
    protected function wrapJsonPath($value)
    {
        return '\'$."'.str_replace('->', '"."', $value).'"\'';
    }
```

### Steps To Reproduce:
* MySQL test
  * `set @json = '{"test": [{"name": "laravel"}, {"name": "symfony"}]}';`
  * run `select json_set(@json, '$.test[0].name', "lumen");`,  no double quotation mark
    * result: `{"test": [{"name": "lumen"}, {"name": "symfony"}]} `, update successfully
  * run `select json_set(@json, '$."test[0]".name', "lumen");`, with double quotation mark
    * result: `{"test": [{"name": "laravel"}, {"name": "symfony"}]} `, update failed
* Laravel test
  *  The sql generate by MySQL grammer: 
~~~~
      update `xx` set `a` = json_set(`a`, '$."test[0]"."name"', lumen)
~~~~
  * This sql has double quotes wrap up `test[0]`, but it will not work for array in MySQL json field. Unless remove the double quotes.

**Comment by staudenmeir** [comment]
I don't think we can support this because of the ambiguity:

```php
->update(['column->foo[0]' => 'bar']);
```

We don't know whether the user wants to update the first array item or they have an object with a key that is actually named `foo[0]`.

The first scenario is probably more likely, but it's more complicated to implement and would be a breaking change.

**Comment by driesvints** [comment]
I'll flag this as a bug for now but I also don't really see how we could solve this soon. Feel free to send in a PR if someone can come up with a non-BC solution.

**Comment by nulele** [comment]
Hi,

I found a similar problem in the 5.8 version.

Consider this JSON string contained in the **data** field of the **heroes** table:
```
[
   {
      "name":"superman",
      "type":"dc",
      "alignment":"good",
   },
   {
      "name":"batman",
      "type":"dc",
      "alignment":"good",
   },
   {
      "name":"joker",
      "type":"dc",
      "alignment":"evil",
   }
]
```
Actually this query cannot be done through Query Builder:
```
SELECT * FROM `heroes` WHERE json_unquote(json_extract(`heroes`.`data`, '$.*."alignment"')) = 'good';
```
Because the *wrapJsonPath* function prints out this (notice quotes around *):
```
SELECT * FROM `heroes` WHERE json_unquote(json_extract(`heroes`.`data`, '$."*"."alignment"')) = 'good';
```
I think the * symbol should be bypassed and not quoted.

**Comment by staudenmeir** [comment]
This would also be a breaking change. An array key could actually be named `[*]`.

**Comment by staudenmeir** [comment]
> Actually this query cannot be done through Query Builder:

> SELECT * FROM `heroes` WHERE json_unquote(json_extract(`heroes`.`data`, '$[*]."alignment"')) = 

This doesn't work for me as a raw query either.

You can use `whereJsonContains()` instead:

```php
Hero::whereJsonContains('data', ['alignment' => 'good'])->get();
```

**Comment by iamgergo** [comment]
Hi,

I bumped into similar issues and I had the following experiences (on MariaDB 10.4.11):

I have a JSON array of objects, named files:

```json
{
    "files": [
        {"name": "file_1", "url": "..."},
        {"name": "file_2", "url": "..."},
    ]
}
```

I called the following update on the model, and it was working as expected:

```php
$model->update(['files->0->name' => 'other_name']);
```

I called an update with an index that was not present, the update converted the array to an object and then inserted the value with the new index – which is logical I think.

```php
$model->update(['files->5->name' => 'new file']);
```

```json
{
    "files": {
        "0": {"name": "file_1", "url": "..."},
        "1": {"name": "file_2", "url": "..."},
        "5": {"name": "new file", "url": "..."},
    }
}
```
> If the array was empty and the key was 0 (files->0->name), the update kept it as an array and inserted the value to 0 index. If the key was other than 0, then it was working like the non-existing indexes/keys.

-----

So, it turns out, the working syntax is not `column->field[0]` but `column->field->0`. Also, with this approach there is no ambiguity issue, because we could even use `column->field[0]->0`. Here it's clear, that we access the subsets of the JSON by using the `->` operator and not the `field[0]` syntax.

Also, this syntax is the same as the one in #31448.

In my opinion, this works as it should. Or do I miss anything here?

**Comment by staudenmeir** [comment]
@iamgergo Your code works because you are updating a single model.

@eleven26 is referring to queries like `Model::where(...)->update(...)`.

When you update a nested JSON value on a single model, Eloquent replaces the value _in PHP_ and the UPDATE query replaces the whole column value.

**Comment by taylorotwell** [comment]
Do we really have any idea what direction we want to go on this? I agree with other commenters that it is likely updating a numerically indexed array value is the vastly more common use case than a string key containing "[0]", etc.

However, either way, we are going to break some applications it seems unless someone has an idea that has not been discussed so far. 

However, If we did change this to interpret the array offset pattern as being intended to represent the numerical offset of an array, we would introduce an inconsistency with way you can update single models using the `update` method (which calls `fill` which presumably calls `Arr::set`). So, we would have to consider if we want to reconcile that new inconsistency. That would introduce a second breaking change.

**Comment by nikosv** [comment]
Searching for an issue i am facing in Laravel 6 with jsonb columns and eloquent where clauses in Postgres i found this issue and while my issue is not related to updates i think it might be relevant (if i need to open a new issue please let me know).

The problem is related to where clauses in numeric indexes. 
And more specifically
->where('data->information->145->value', 'whatever');

Generates:
WHERE "data"->'information'->145->>'value' = 'whatever'

Intead of:
 WHERE "data"->'information'->'145'->>'value' = 'whatever'

A workaround would be to whereJsonContains like so:
->whereJsonContains('data->information', ['145' => ['value' => 'whatever']])

**Comment by ollieread** [comment]
Could you not make use of a custom `Illuminate\Database\Query\Expression` class?

Examples:

```php
$something->update([
    JsonPath::make('foo')->index(0)->key('name') => 'bar',
]);
```

Using `ArrayAccess` and magic methods you could even do something like:

```php
$something->update([
    JsonPath::make()->foo[0]->name => 'bar',
]);

// Or

$something->update([
    JsonPath::make()->foo->all()->name => 'bar',
]);
```

It's not the nicest solution, though it does feel nicer than doing `'foo[0]->name'`, could help prevent mistakes in using a string.

**Comment by taylorotwell** [comment]
@ollieread that's a pretty good solution actually. Would you have time to PR it so we can take a look?

**Comment by ollieread** [comment]
@taylorotwell happily. Do you have a preference for which of the methods I suggested?

**Comment by taylorotwell** [comment]
For future reference, a PR was started here: https://github.com/laravel/framework/pull/34515

However, it still needs completion and support / testing for multiple database drivers if anyone wants to pick it up and solve this issue.

**Comment by derekmd** [comment]
Those `JsonPath::make()` examples look nice and clean but PHP associative array keys cannot be objects, even when the `__toString()` magic method exists. So that `update()` syntax can't be pursued.

### Supporting the `JsonPath` object

I picked up that dev branch and updated the new `JsonPath` class to rename MySQL/Oracle jargon (...cell? array index) to language better understood by devs across each of Laravel's database drivers. It also supports the full range of JSON path expressions such as basic example `'$**.b'` found in the [MySQL JSON docs](https://dev.mysql.com/doc/refman/8.0/en/json.html#json-path-syntax) for `SELECT JSON_EXTRACT('{"a": {"b": 1}, "c": {"b": 2}}', '$**.b');`.

Branch https://github.com/laravel/framework/compare/8.x...derekmd:json-path-expression allows object-built JSON path expressions to be done on WHERE clauses. e.g., from this issue's root post:

```php
DB::table('xx')->where(JsonPath::make('a')->test[0]->name, 'lumen')->get();
```

```sql
SELECT * from `xx` WHERE JSON_UNQUOTE(JSON_EXTRACT(`a`, $."test"[0]."name")) = 'lumen';
```

### Changes required to `Builder@update()`

To make the query builder `update()` method support  `JsonPath` as an argument, the method could (hack-ily) be updated to support argument tuples?

```php
DB::table('xx')->update([
    [JsonPath::make('a')->test[0]->name, 'lumen'],
    [JsonPath::make('a')->test[0]->version, 8.3],
    'b' => 'c',
]);
```

```sql
UPDATE `xx`
SET `a` = JSON_SET(`a`, $."test"[0]."name", 'lumen', $."test"[0]."name", 8.3),
    `b` = 'c';
```

That really doesn't look and feel like a Laravel-ish solution though. Otherwise a new method `Illuminate\Database\Query\Builder@updateJsonPath()` is required with breaking changes to:
* `Illuminate\Database\Query\Grammars\Grammar@compileUpdate()`
* `Illuminate\Database\Query\Grammars\Grammar@compileUpdateColumns()`.

### Other solutions to update a single nested JSON field

Although for this whole issue, when `Builder@update()` can't do the job I'd argue:

* Run a raw SQL query if an atomic JSON column update is important. How often does even the most atypical Laravel app update the database for a single JSON array index? This issue has been opened 2+ years with little demand to get this case covered. (Start your yelling now.)
  * Caveat: Laravel devs may also not be realizing they're updating/selecting a JSON object key and not an array index. Since this doesn't throw an exception, there may be some inaccurate `SELECT` report queries out there missing the (original not updated) rows containing a JSON array index.
* You can also instead pull JSON database rows into PHP (Eloquent model attribute), mutate changes, and then run the SQL to update the whole JSON document in the database
  * Assuming another update isn't running in parallel which may cause partial data loss from an update overwriting another that just occurred. Maybe don't do this.
* if nested `UPDATE`s (or common `WHERE` queries) are being done on a JSON column, it may make more sense as a plain normalized database column? Or another one-to-many table for an array of items? I've personally done this before on Stripe API response body JSON to pull out a charge ID that could be in JSON path `$.id` (payment attempt response) or `$.charge` (refund response). A single `charge_id` text column will do for querying payments & related refunds.

**Comment by ollieread** [comment]
I find myself with additional free time now, is this still something that needs looking into?

**Comment by driesvints** [comment]
@ollieread taylor is gonna look at Derek's reply soon so we can pick one of those solutions. To me nr 2 seems the wanted one. What's your take?

**Comment by derekmd** [comment]
Yeah, I planned to eventually submit solution 2 since the code change is simple and it covers most JSON path queries. Since that must be submitted to 9.x (Jan 2022) I was waiting to see if CI test coverage for MS SQL + Postgres was going to be added so I can get JSON queries running for each driver in the test suite. Or I can submit it now and tick of this issue checkbox off since I know that MS SQL dev branch is test suite refactor hell.

**Comment by ollieread** [comment]
I think they're all viable solutions @driesvints. If i had to pick one, I'd go with 2, but I also understand the ramifications of the breaking changes.

Good solutions @derekmd , thanks for picking this up where I ran out of free time :p

**Comment by driesvints** [comment]
@derekmd i think it's best that you submit it as I don't know when I'll get to finishing up the postgres and mssql builds.

**Comment by driesvints** [comment]
Whohoo, glad we got this one closed. Thanks @derekmd! 🥳
